crates/iceberg/src/arrow.rs

```diff
@@ -106,3 +114,732 @@ impl ArrowReader {
+
+                let key_result = {
+                    visitor.before_map_key(key_field)?;
+                    let ret = visit_type(key_field.data_type(), visitor)?;
+                    visitor.after_map_key(key_field)?;
+                    ret
+                };
+
+                let value_result = {
+                    visitor.before_map_value(value_field)?;
+                    let ret = visit_type(value_field.data_type(), visitor)?;
+                    visitor.after_map_value(value_field)?;
+                    ret
+                };
+
+                visitor.map(r#type, key_result, value_result)
+            }
+            _ => Err(Error::new(
+                ErrorKind::DataInvalid,
+                "Map field must have struct type",
+            )),
+        },
+        DataType::Struct(fields) => visit_struct(fields, visitor),
+        other => Err(Error::new(
+            ErrorKind::DataInvalid,
+            format!("Cannot visit Arrow data type: {other}"),
+        )),
+    }
+}
+
+/// Visit list types in post order.
+#[allow(dead_code)]
+fn visit_list<V: ArrowSchemaVisitor>(
+    data_type: &DataType,
+    element_field: &Field,
+    visitor: &mut V,
+) -> Result<V::T> {
+    visitor.before_list_element(element_field)?;
+    let value = visit_type(element_field.data_type(), visitor)?;
+    visitor.after_list_element(element_field)?;
+    visitor.list(data_type, value)
+}
+
+/// Visit struct type in post order.
+#[allow(dead_code)]
+fn visit_struct<V: ArrowSchemaVisitor>(fields: &Fields, visitor: &mut V) -> Result<V::T> {
+    let mut results = Vec::with_capacity(fields.len());
+    for field in fields {
+        visitor.before_field(field)?;
+        let result = visit_type(field.data_type(), visitor)?;
+        visitor.after_field(field)?;
+        results.push(result);
+    }
+
+    visitor.r#struct(fields, results)
+}
+
+/// Visit schema in post order.
+#[allow(dead_code)]
+fn visit_schema<V: ArrowSchemaVisitor>(schema: &ArrowSchema, visitor: &mut V) -> Result<V::U> {
+    let mut results = Vec::with_capacity(schema.fields().len());
+    for field in schema.fields() {
+        visitor.before_field(field)?;
+        let result = visit_type(field.data_type(), visitor)?;
+        visitor.after_field(field)?;
+        results.push(result);
+    }
+    visitor.schema(schema, results)
+}
+
+/// Convert Arrow schema to ceberg schema.
+#[allow(dead_code)]
+pub fn arrow_schema_to_schema(schema: &ArrowSchema) -> Result<Schema> {
+    let mut visitor = ArrowSchemaConverter::new();
+    visit_schema(schema, &mut visitor)
+}
+
+const ARROW_FIELD_ID_KEY: &str = "PARQUET:field_id";
+const ARROW_FIELD_DOC_KEY: &str = "doc";
+
+fn get_field_id(field: &Field) -> Result<i32> {
+    if let Some(value) = field.metadata().get(ARROW_FIELD_ID_KEY) {
+        return value.parse::<i32>().map_err(|e| {
+            Error::new(
+                ErrorKind::DataInvalid,
+                format!("Failed to parse field id: {e}"),
+            )
```

Comment:
```suggestion
            Error::new(
                ErrorKind::DataInvalid,
                format!("Failed to parse field id"),
            ).with_context("value", value)
            .with_source(e)
```

This will make error reporting easier to read.